tests/estimators/certification/test_smooth_vit.py

```diff
@@ -0,0 +1,331 @@
+                    super().__init__()
+                    self.avg_pool = torch.nn.AvgPool2d(patch_size)
+
+                def forward(self, ones_mask):
+                    B = ones_mask.shape[0]
+                    ones_mask = ones_mask[0].unsqueeze(0)  # take the first mask
+                    ones_mask = self.avg_pool(ones_mask)[0]
+                    ones_mask = torch.where(ones_mask.view(-1) > 0)[0] + 1
+                    ones_mask = torch.cat([torch.IntTensor(1).fill_(0), ones_mask]).unsqueeze(0)
+                    ones_mask = ones_mask.expand(B, -1)
+                    return ones_mask
+
+            mask_processor = MaskProcessor()
+            patch_mask = mask_processor(mask)
+
+            # x = self.pos_drop(x) # B, N, C
+            if patch_mask is not None:
+                # patch_mask is B, K
+                B, N, C = x.shape
+                if len(patch_mask.shape) == 1:  # not a separate one per batch
+                    x = x[:, patch_mask]
+                else:
+                    patch_mask = patch_mask.unsqueeze(-1).expand(-1, -1, C)
+                    x = torch.gather(x, 1, patch_mask)
+            return x
+
+        @classmethod
+        def embedder(cls, x, pos_embed, cls_token):
+            """
+            NB, original code used the pos embed from the divit rather than vit
+            (which we pull from our model) which we use here.
+            """
+            x = torch.cat((cls_token.expand(x.shape[0], -1, -1), x), dim=1)
+            return x + pos_embed
+
+    def forward_features(self, x: torch.Tensor) -> torch.Tensor:
+        """
+        This is a copy of the function in ArtViT.forward_features
+        except we also perform an equivalence assertion compared to the implementation
+        in https://github.com/MadryLab/smoothed-vit (see MadrylabImplementations class above)
+
+        The forward pass of the ViT.
+
+        :param x: Input data.
+        :return: The input processed by the ViT backbone
+        """
+        import copy
+
+        ablated_input = False
+        if x.shape[1] == self.in_chans + 1:
+            ablated_input = True
+
+        if ablated_input:
+            x, ablation_mask = x[:, : self.in_chans], x[:, self.in_chans : self.in_chans + 1]
+
+        x = self.patch_embed(x)
+
+        madry_embed = MadrylabImplementations.embedder(copy.copy(x), self.pos_embed, self.cls_token)
+        x = self._pos_embed(x)
+        assert torch.equal(madry_embed, x)
+
+        # pass the x into the token dropping code from ...
+        madry_dropped = MadrylabImplementations.token_dropper(copy.copy(x), ablation_mask)
+
+        if self.to_drop_tokens and ablated_input:
+            ones = self.ablation_mask_embedder(ablation_mask)
+            to_drop = torch.sum(ones, dim=2)
+            indexes = torch.gt(torch.where(to_drop > 1, 1, 0), 0)
+            x = self.drop_tokens(x, indexes)
+
+        print(torch.equal(madry_dropped, x))
```

Comment:
Please use logger.

Reply:
line was redundant so simply removing it